cmd/pod-scaler/producer.go

```diff
@@ -0,0 +1,482 @@
+		Container: string(metric[LabelNameContainer]),
+	}
+}
+
+// uncoveredRanges determines the largest subset ranges of r that are not covered by
+// existing data in the querier.
+func (q *querier) uncoveredRanges(cluster string, r TimeRange) []TimeRange {
+	q.lock.RLock()
+	defer q.lock.RUnlock()
+	return uncoveredRanges(r, q.data.RangesByCluster[cluster])
+}
+
+func uncoveredRanges(r TimeRange, coverage []TimeRange) []TimeRange {
+	var covered []TimeRange
+	for _, extent := range coverage {
+		startsInside := within(extent.Start, r)
+		endsInside := within(extent.End, r)
+		switch {
+		case startsInside && endsInside:
+			covered = append(covered, extent)
+		case startsInside && !endsInside:
+			covered = append(covered, TimeRange{
+				Start: extent.Start,
+				End:   r.End,
+			})
+		case !startsInside && endsInside:
+			covered = append(covered, TimeRange{
+				Start: r.Start,
+				End:   extent.End,
+			})
+		case extent.Start.Before(r.Start) && extent.End.After(r.End):
+			covered = append(covered, TimeRange{
+				Start: r.Start,
+				End:   r.End,
+			})
+		}
+	}
+	sort.Slice(covered, func(i, j int) bool {
+		return covered[i].Start.Before(covered[j].Start)
+	})
+	covered = coalesce(covered)
+
+	if len(covered) == 0 {
+		return []TimeRange{r}
+	}
+	var uncovered []TimeRange
+	if !covered[0].Start.Equal(r.Start) {
+		uncovered = append(uncovered, TimeRange{Start: r.Start, End: covered[0].Start})
+	}
+	for i := 0; i < len(covered)-1; i++ {
+		uncovered = append(uncovered, TimeRange{Start: covered[i].End, End: covered[i+1].Start})
+	}
+	if !covered[len(covered)-1].End.Equal(r.End) {
+		uncovered = append(uncovered, TimeRange{Start: covered[len(covered)-1].End, End: r.End})
+	}
+	return uncovered
+}
+
+// within determines if the time falls within the range
+func within(t time.Time, r TimeRange) bool {
+	return (r.Start.Equal(t) || r.Start.Before(t)) && (r.End.Equal(t) || r.End.After(t))
+}
+
+// coalesce minimizes the number of timeRanges that are needed to describe a set of times.
+// The output is sorted by start time of the remaining ranges.
+func coalesce(input []TimeRange) []TimeRange {
+	for {
+		coalesced := coalesceOnce(input)
+		if len(coalesced) == len(input) {
+			sort.Slice(coalesced, func(i, j int) bool {
+				return coalesced[i].Start.Before(coalesced[j].Start)
+			})
+			return coalesced
+		}
+		input = coalesced
+	}
+}
+
+func coalesceOnce(input []TimeRange) []TimeRange {
+	for i := 0; i < len(input); i++ {
+		for j := i; j < len(input); j++ {
+			var coalesced *TimeRange
+			if input[i].End.Equal(input[j].Start) {
+				coalesced = &TimeRange{
+					Start: input[i].Start,
+					End:   input[j].End,
+				}
+			}
+			if input[i].Start.Equal(input[j].End) {
+				coalesced = &TimeRange{
+					Start: input[j].Start,
+					End:   input[i].End,
+				}
+			}
+			if coalesced != nil {
+				return append(input[:i], append(input[i+1:j], append(input[j+1:], *coalesced)...)...)
+			}
+		}
+	}
+	return input
+}
```

Comment:
If you didn't make an off-by-one, unwanted overlap, out of bounds or something similar somewhere here I'll buy you a beer next time we meet.

Reply:
There are tests! It works!